### PR TITLE
Add `CoupledTrainStepper`

### PR DESCRIPTION
**Pull request by jpdunc23**
Adds `train_stepper: CoupledTrainStepperConfig` to the coupled training config, which configures and builds a  `CoupledTrainStepper` implementing `TrainStepperABC`.

**WARNING**: This is a breaking change for existing coupled training configs.

Changes:
- Component stepper `loss: StepLossConfig` and `loss_contributions: LossContributionsConfig` are now configured via the `ocean: ComponentTrainingConfig` and `atmosphere: ComponentTrainingConfig` attributes of  `CoupledTrainStepperConfig`.
- `CoupledStepper` no longer implements `TrainStepperABC`.
- Removed public `loss_obj` and `effective_loss_scaling` properties from `fme.ace.stepper.Stepper` and added a new public method `build_loss`.

- [x] Tests added

### CODE REVIEW

File: fme/coupled/loss.py [hunk 1]
```diff
@@ -57,11 +61,11 @@ class LossContributionsConfig:
 
     def build(
         self,
-        loss_obj: Callable[[TensorMapping, TensorMapping, int], torch.Tensor],
+        loss_obj: StepLoss,
         time_dim: int,
     ) -> StepLossABC:
         if self.n_steps == 0 or self.weight == 0.0:
-            return NullLossContributions()
+            return NullLossContributions(loss_obj)
```
Comment (jpdunc23) [inline fme/coupled/loss.py:68]: This preserves the existing behavior where we used a component stepper's `effective_loss_scaling` to compute `mse_fractional_components` metrics even if the stepper had no loss contribution in coupled training.

File: fme/coupled/test_loss.py [hunk 2]
```diff
@@ -60,6 +61,10 @@ def __init__(
         self._loss_obj = loss_obj
         self._time_dim = time_dim
 
+    @property
+    def effective_loss_scaling(self):
+        raise NotImplementedError
```
Comment (mcgibbon) [inline fme/coupled/test_loss.py]:
```suggestion
        raise NotImplementedError()
```

File: fme/coupled/test_loss.py [hunk 3]
```diff
@@ -153,13 +159,12 @@ def test_null_loss_contributions(steps_thru_atmos_7, ocean_config_kwargs):
     # test LossContributionsConfig with n_steps = 0
     atmos_loss_config = LossContributionsConfig()
     atmosphere_loss = atmos_loss_config.build(
-        loss_obj=lambda *_, **__: torch.tensor(5.25),
+        loss_obj=Mock(spec=StepLoss, side_effect=lambda *_, **__: torch.tensor(5.25)),
```
Comment (mcgibbon) [inline fme/coupled/test_loss.py]: nit: The three lines changed use three different ways to specify the loss side-effect - via mae_loss, via a lambda function returning a constant, and via a return_value instead of a side_effect. You could consider using return_value for this one to reduce that down to 2 ways, at least.

File: fme/coupled/test_stepper.py [hunk 4]
```diff
@@ -1501,7 +1503,12 @@ def test_train_on_batch_with_derived_variables():
         n_forward_times_atmosphere=2,
         n_samples=3,
     )
-    output = coupler.train_on_batch(
+    train_stepper_config = CoupledTrainStepperConfig(
```
Comment (mcgibbon) [inline fme/coupled/test_stepper.py:1506]: nit: Avoid the 3x copy-paste of this process by making a `get_train_stepper_and_batch` helper that does it and calls `get_stepper_and_batch` internally.

File: fme/coupled/stepper.py [hunk 5]
```diff
@@ -1319,12 +1290,169 @@ def predict(
...
+@dataclasses.dataclass
+class ComponentTrainingConfig:
+    loss: StepLossConfig
+    loss_contributions: LossContributionsConfig = dataclasses.field(
+        default_factory=lambda: LossContributionsConfig()
+    )
+
+
+@dataclasses.dataclass
+class CoupledTrainStepperConfig:
```
Comment (mcgibbon) [inline fme/coupled/stepper.py:1333]: Do you have an example of the updated training config committed somewhere I could check out? It would be nice to have a baseline config for coupled training, if so I could see the changes to the baseline in this PR.

Update: Ah I see test_train.py mostly fits this purpose, good. Still, could be nice to have a baseline in the future.

File: fme/coupled/test_train.py [hunk 6]
```diff
@@ -109,8 +107,6 @@
           out_names: {ocean_out_names}
   atmosphere:
     timedelta: 1D
-    loss_contributions:
-      n_steps: {loss_atmos_n_steps}
     stepper:
       loss:
```
Comment (mcgibbon) [inline fme/coupled/test_train.py]: Question: Why is `loss: type: MSE` in both the `atmosphere: stepper:` and in the `train_stepper: atmosphere:`? I am guessing because we haven't updated the ACE configs yet and it's required in the config, in which case that's fine, but I though I should ask to be sure.

File: fme/coupled/train/train.py [hunk 7]
```diff
@@ -58,8 +58,8 @@ def build_trainer(builder: TrainBuilders, config: TrainConfig) -> Trainer:
         n_timesteps_atmosphere=n_timesteps_atmosphere,
         ocean_normalize=stepper.ocean.normalizer.normalize,
         atmosphere_normalize=stepper.atmosphere.normalizer.normalize,
-        ocean_loss_scaling=stepper.ocean.effective_loss_scaling,
-        atmosphere_loss_scaling=stepper.atmosphere.effective_loss_scaling,
+        ocean_loss_scaling=stepper.effective_loss_scaling.ocean,
```
Comment (mcgibbon) [inline fme/coupled/train/train.py]: nit: pass `loss_scaling: stepper.loss_scaling` instead of two arguments containing the parts

File: fme/coupled/test_stepper.py (hunk 4 above)
Comment (jpdunc23) [reply fme/coupled/test_stepper.py:1506]: Good idea, but I'll defer this cleanup to https://github.com/ai2cm/ace/pull/814 since the way in which the train stepper is built is going to change.

File: fme/coupled/stepper.py (hunk 5 above)
Comment (jpdunc23) [reply fme/coupled/stepper.py:1333]: Agreed, I will work on a new PR to add the baseline.

File: fme/coupled/test_train.py (hunk 6 above)
Comment (jpdunc23) [reply fme/coupled/test_train.py]: That's right, although including it in the yaml here isn't strictly necessary since there is a default value on `StepperConfig`. I'll remove it so it's a bit clearer here.
